File: backend/shared-infrastructure/knowledge-base-services/kb-6-formulary/internal/middleware/middleware.go
package middleware

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"runtime/debug"
	"sync"
	"time"
)

// Middleware represents a middleware function
type Middleware func(http.Handler) http.Handler

// Chain applies multiple middleware functions in sequence
func Chain(h http.Handler, middlewares ...Middleware) http.Handler {
	for i := len(middlewares) - 1; i >= 0; i-- {
		h = middlewares[i](h)
	}
	return h
}

// RequestLogging logs HTTP requests and responses
func RequestLogging() Middleware {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			start := time.Now()
			
			// Wrap the response writer to capture status
			wrapped := &responseWriter{ResponseWriter: w, statusCode: 200}
			
			next.ServeHTTP(wrapped, r)
			
			duration := time.Since(start)
			log.Printf("HTTP %s %s - %d - %v", r.Method, r.URL.Path, wrapped.statusCode, duration)
		})
	}
}

// CORS adds Cross-Origin Resource Sharing headers
func CORS() Middleware {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Access-Control-Allow-Origin", "*")
			w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
			w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization, X-Requested-With")
			w.Header().Set("Access-Control-Max-Age", "86400")
			
			if r.Method == "OPTIONS" {
				w.WriteHeader(http.StatusNoContent)
				return
			}
			
			next.ServeHTTP(w, r)
		})
	}
}

// RateLimit implements basic rate limiting per IP
func RateLimit() Middleware {
	type client struct {
		requests int
		lastSeen time.Time
	}
	
	clients := make(map[string]*client)
	mu := sync.Mutex{}
	
	// Clean up old entries every minute
	go func() {
		for {
			time.Sleep(1 * time.Minute)
			mu.Lock()
			now := time.Now()
			for ip, c := range clients {
				if now.Sub(c.lastSeen) > 1*time.Minute {
					delete(clients, ip)
				}
			}
			mu.Unlock()
		}
	}()
	
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ip := getClientIP(r)
			
			mu.Lock()
			c, exists := clients[ip]
			if !exists {
				c = &client{requests: 0, lastSeen: time.Now()}
				clients[ip] = c
			}
			
			// Reset counter if more than a minute has passed
			if time.Since(c.lastSeen) > 1*time.Minute {
				c.requests = 0
			}
			
			c.requests++
			c.lastSeen = time.Now()
			
			// Rate limit: 100 requests per minute
			if c.requests > 100 {
				mu.Unlock()
				http.Error(w, "Rate limit exceeded", http.StatusTooManyRequests)
				return
			}
			mu.Unlock()
			
			next.ServeHTTP(w, r)
		})
	}
}

// RequestTimeout adds a timeout to requests
func RequestTimeout(timeout time.Duration) Middleware {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ctx, cancel := context.WithTimeout(r.Context(), timeout)
			defer cancel()
			
			r = r.WithContext(ctx)
			next.ServeHTTP(w, r)
		})
	}
}

// Recovery recovers from panics and returns a 500 error
func Recovery() Middleware {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			defer func() {
				if err := recover(); err != nil {
					log.Printf("Panic recovered: %v\n%s", err, debug.Stack())
					
					w.Header().Set("Content-Type", "application/json")
					w.WriteHeader(http.StatusInternalServerError)
					
					errorResponse := map[string]interface{}{
						"error":     "Internal server error",
						"timestamp": time.Now().UTC(),
						"path":      r.URL.Path,
					}
					
					json.NewEncoder(w).Encode(errorResponse)
				}
			}()
			
			next.ServeHTTP(w, r)
		})
	}
}

// MetricsHandler provides a basic metrics endpoint
func MetricsHandler() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		metrics := map[string]interface{}{
			"service":        "kb-6-formulary",
			"version":        "1.0.0",
			"uptime_seconds": time.Since(startTime).Seconds(),
			"timestamp":      time.Now().UTC(),
			"go_version":     "go1.21+",
			"status":         "healthy",
		}
		
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(metrics)
	}
}

// Authentication middleware for Bearer token validation
func Authentication() Middleware {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Skip authentication for health checks and public endpoints
			if r.URL.Path == "/health" || r.URL.Path == "/metrics" || r.URL.Path == "/" {
				next.ServeHTTP(w, r)
				return
			}
			
			authHeader := r.Header.Get("Authorization")
			if authHeader == "" {
				http.Error(w, "Authorization header required", http.StatusUnauthorized)
				return
			}
			
			// Basic Bearer token validation (in production, integrate with auth service)
			if len(authHeader) < 7 || authHeader[:7] != "Bearer " {
				http.Error(w, "Invalid authorization header format", http.StatusUnauthorized)
				return
			}
			
			token := authHeader[7:]
			if !isValidToken(token) {
				http.Error(w, "Invalid or expired token", http.StatusUnauthorized)
				return
			}
			
			next.ServeHTTP(w, r)
		})
	}
}

// responseWriter wraps http.ResponseWriter to capture status code
type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

// getClientIP extracts client IP from request
func getClientIP(r *http.Request) string {
	// Check X-Forwarded-For header first
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		return xff
	}
	
	// Check X-Real-IP header
	if xri := r.Header.Get("X-Real-IP"); xri != "" {
		return xri
	}
	
	// Fall back to RemoteAddr
	return r.RemoteAddr
}

// isValidToken validates Bearer tokens (placeholder implementation)
func isValidToken(token string) bool {
	// In production, this would validate against your auth service
	// For now, accept any non-empty token longer than 10 characters
	return len(token) > 10
}

// startTime tracks when the service started for uptime metrics
var startTime = time.Now()

// RequestID middleware adds a unique request ID to each request
func RequestID() Middleware {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			requestID := generateRequestID()
			w.Header().Set("X-Request-ID", requestID)
			
			// Add to context for use in handlers
			ctx := context.WithValue(r.Context(), "request_id", requestID)
			r = r.WithContext(ctx)
			
			next.ServeHTTP(w, r)
		})
	}
}

// generateRequestID creates a simple request ID
func generateRequestID() string {
	return fmt.Sprintf("kb6-%d", time.Now().UnixNano())
}

// Security headers middleware
func SecurityHeaders() Middleware {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("X-Content-Type-Options", "nosniff")
			w.Header().Set("X-Frame-Options", "DENY")
			w.Header().Set("X-XSS-Protection", "1; mode=block")
			w.Header().Set("Referrer-Policy", "strict-origin-when-cross-origin")
			
			next.ServeHTTP(w, r)
		})
	}
}

File: backend/services/medication-service/flow2-rust-engine/src/api/middleware.rs
//! Production-grade middleware for the REST API server

use axum::{
    extract::{Request, State},
    http::{HeaderMap, StatusCode},
    middleware::Next,
    response::Response,
    Json,
};
use crate::api::server::ApiState;
use std::time::Instant;
use tracing::{info, warn, error, Span};
use uuid::Uuid;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use tokio::time::{Duration, sleep};

/// Request tracking middleware
pub async fn request_tracking_middleware(
    mut request: Request,
    next: Next,
) -> Result<Response, StatusCode> {
    let start_time = Instant::now();
    
    // Generate or extract request ID
    let request_id = request
        .headers()
        .get("x-request-id")
        .and_then(|h| h.to_str().ok())
        .unwrap_or(&Uuid::new_v4().to_string())
        .to_string();

    // Add request ID to headers for downstream services
    request.headers_mut().insert(
        "x-request-id",
        request_id.parse().unwrap(),
    );

    // Create tracing span with request context
    let span = tracing::info_span!(
        "api_request",
        request_id = %request_id,
        method = %request.method(),
        uri = %request.uri(),
    );

    let _enter = span.enter();

    info!("📥 Request started: {} {}", request.method(), request.uri());

    // Process request
    let response = next.run(request).await;

    let duration = start_time.elapsed();
    let status = response.status();

    // Log completion
    if status.is_success() {
        info!("✅ Request completed: {} ({:?})", status, duration);
    } else if status.is_client_error() {
        warn!("⚠️  Request failed (client error): {} ({:?})", status, duration);
    } else {
        error!("❌ Request failed (server error): {} ({:?})", status, duration);
    }

    Ok(response)
}

/// Rate limiting middleware
#[derive(Clone)]
pub struct RateLimiter {
    requests: Arc<Mutex<HashMap<String, Vec<Instant>>>>,
    max_requests: usize,
    window_duration: Duration,
}

impl RateLimiter {
    /// Create a new rate limiter
    pub fn new(max_requests: usize, window_duration: Duration) -> Self {
        Self {
            requests: Arc::new(Mutex::new(HashMap::new())),
            max_requests,
            window_duration,
        }
    }

    /// Check if request is allowed
    pub fn is_allowed(&self, client_id: &str) -> bool {
        let mut requests = self.requests.lock().unwrap();
        let now = Instant::now();
        
        // Get or create request history for this client
        let client_requests = requests.entry(client_id.to_string()).or_insert_with(Vec::new);
        
        // Remove old requests outside the window
        client_requests.retain(|&request_time| {
            now.duration_since(request_time) < self.window_duration
        });
        
        // Check if under limit
        if client_requests.len() < self.max_requests {
            client_requests.push(now);
            true
        } else {
            false
        }
    }

    /// Clean up old entries periodically
    pub async fn cleanup_task(&self) {
        let mut interval = tokio::time::interval(Duration::from_secs(60));
        
        loop {
            interval.tick().await;
            
            let mut requests = self.requests.lock().unwrap();
            let now = Instant::now();
            
            // Remove clients with no recent requests
            requests.retain(|_, client_requests| {
                client_requests.retain(|&request_time| {
                    now.duration_since(request_time) < self.window_duration * 2
                });
                !client_requests.is_empty()
            });
        }
    }
}

/// Rate limiting middleware function
pub async fn rate_limiting_middleware(
    headers: HeaderMap,
    State(rate_limiter): State<RateLimiter>,
    request: Request,
    next: Next,
) -> Result<Response, (StatusCode, Json<serde_json::Value>)> {
    // Extract client identifier (IP address or API key)
    let client_id = headers
        .get("x-forwarded-for")
        .or_else(|| headers.get("x-real-ip"))
        .and_then(|h| h.to_str().ok())
        .unwrap_or("unknown")
        .to_string();

    if !rate_limiter.is_allowed(&client_id) {
        warn!("🚫 Rate limit exceeded for client: {}", client_id);
        return Err((
            StatusCode::TOO_MANY_REQUESTS,
            Json(serde_json::json!({
                "error": "Rate limit exceeded",
                "message": "Too many requests. Please try again later.",
                "retry_after": "60 seconds"
            }))
        ));
    }

    Ok(next.run(request).await)
}

/// Authentication middleware
pub async fn auth_middleware(
    State(state): State<ApiState>,
    headers: HeaderMap,
    request: Request,
    next: Next,
) -> Result<Response, (StatusCode, Json<serde_json::Value>)> {
    // Skip auth for health endpoints
    let path = request.uri().path();
    if path.starts_with("/health") || path.starts_with("/metrics") || path.starts_with("/status") {
        return Ok(next.run(request).await);
    }

    // Check if authentication is enabled in configuration
    if !state.config.security.enable_auth {
        // Authentication is disabled, allow all requests
        return Ok(next.run(request).await);
    }

    // Check for API key or authorization header
    let auth_header = headers
        .get("authorization")
        .or_else(|| headers.get("x-api-key"))
        .and_then(|h| h.to_str().ok());

    match auth_header {
        Some(token) => {
            // Validate token against configured API keys
            if state.config.security.api_keys.contains(&token.to_string()) || validate_token(token) {
                info!("🔐 Authentication successful");
                Ok(next.run(request).await)
            } else {
                warn!("🚫 Authentication failed: invalid token");
                Err((
                    StatusCode::UNAUTHORIZED,
                    Json(serde_json::json!({
                        "error": "Authentication failed",
                        "message": "Invalid or expired token"
                    }))
                ))
            }
        }
        None => {
            warn!("🚫 Authentication failed: missing token");
            Err((
                StatusCode::UNAUTHORIZED,
                Json(serde_json::json!({
                    "error": "Authentication required",
                    "message": "Missing authorization header or API key"
                }))
            ))
        }
    }
}

/// Validate authentication token (simplified implementation)
fn validate_token(token: &str) -> bool {
    // In production, this would validate against a proper auth service
    // For demo purposes, accept any token that starts with "Bearer " or "ApiKey "
    token.starts_with("Bearer ") || token.starts_with("ApiKey ") || token == "development-token"
}

/// CORS middleware
pub async fn cors_middleware(
    request: Request,
    next: Next,
) -> Result<Response, StatusCode> {
    let mut response = next.run(request).await;
    
    let headers = response.headers_mut();
    headers.insert("Access-Control-Allow-Origin", "*".parse().unwrap());
    headers.insert("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS".parse().unwrap());
    headers.insert("Access-Control-Allow-Headers", "Content-Type, Authorization, X-API-Key, X-Request-ID".parse().unwrap());
    headers.insert("Access-Control-Max-Age", "86400".parse().unwrap());
    
    Ok(response)
}

/// Security headers middleware
pub async fn security_headers_middleware(
    request: Request,
    next: Next,
) -> Result<Response, StatusCode> {
    let mut response = next.run(request).await;
    
    let headers = response.headers_mut();
    
    // Security headers
    headers.insert("X-Content-Type-Options", "nosniff".parse().unwrap());
    headers.insert("X-Frame-Options", "DENY".parse().unwrap());
    headers.insert("X-XSS-Protection", "1; mode=block".parse().unwrap());
    headers.insert("Referrer-Policy", "strict-origin-when-cross-origin".parse().unwrap());
    headers.insert("Content-Security-Policy", "default-src 'self'".parse().unwrap());
    
    // API-specific headers
    headers.insert("X-API-Version", "2.0.0".parse().unwrap());
    headers.insert("X-Engine", "rust-recipe-engine".parse().unwrap());
    
    Ok(response)
}

/// Request validation middleware
pub async fn request_validation_middleware(
    headers: HeaderMap,
    request: Request,
    next: Next,
) -> Result<Response, (StatusCode, Json<serde_json::Value>)> {
    // Validate Content-Type for POST requests
    if request.method() == "POST" {
        let content_type = headers
            .get("content-type")
            .and_then(|h| h.to_str().ok())
            .unwrap_or("");

        if !content_type.starts_with("application/json") {
            return Err((
                StatusCode::BAD_REQUEST,
                Json(serde_json::json!({
                    "error": "Invalid Content-Type",
                    "message": "Content-Type must be application/json for POST requests"
                }))
            ));
        }
    }

    // Validate request size (prevent large payloads)
    if let Some(content_length) = headers.get("content-length") {
        if let Ok(length_str) = content_length.to_str() {
            if let Ok(length) = length_str.parse::<usize>() {
                const MAX_PAYLOAD_SIZE: usize = 10 * 1024 * 1024; // 10MB
                if length > MAX_PAYLOAD_SIZE {
                    return Err((
                        StatusCode::PAYLOAD_TOO_LARGE,
                        Json(serde_json::json!({
                            "error": "Payload too large",
                            "message": format!("Request payload exceeds maximum size of {} bytes", MAX_PAYLOAD_SIZE)
                        }))
                    ));
                }
            }
        }
    }

    Ok(next.run(request).await)
}

/// Timeout middleware
pub async fn timeout_middleware(
    request: Request,
    next: Next,
) -> Result<Response, (StatusCode, Json<serde_json::Value>)> {
    const REQUEST_TIMEOUT: Duration = Duration::from_secs(30);
    
    match tokio::time::timeout(REQUEST_TIMEOUT, next.run(request)).await {
        Ok(response) => Ok(response),
        Err(_) => {
            error!("⏰ Request timeout after {:?}", REQUEST_TIMEOUT);
            Err((
                StatusCode::REQUEST_TIMEOUT,
                Json(serde_json::json!({
                    "error": "Request timeout",
                    "message": "Request took too long to process"
                }))
            ))
        }
    }
}
